### ISSUE TITLE
authorize: fallback to using a user's id_token's group claim

### ISSUE THREAD

**Issue by desimone** [pomerium/pomerium#1112]
**Is your feature request related to a problem? Please describe.**

In v0.10.0 groups (and other associate user data) will be stored in the databroker, and group membership will be polled from a given identity provider. The issue is not all identity providers have (or easily expose) their groups endpoint. We should continue to support those OIDC providers that do allow setting groups as part of the id_token claims. 

**Describe the solution you'd like**

Authorize should use the user's identity token's `group` claim as a fallback for groups if available. 

**Describe alternatives you've considered**

**Explain any additional use-cases**

If there are any use-cases that would help us understand the use/need/value please share them as they can help us decide on acceptance and prioritization.

**Additional context**

https://github.com/pomerium/pomerium/pull/885#issuecomment-661087840

**Comment by sly-net** [comment]
Thanks for opening this issue.

In my case, I'm setting up a fully self-hosted solution for Identity and Access management. No Identity as a Service. I'm trying to combine Traefik, Pomerium (in forward-auth mode) and Dex which has a connector to an OpenLDAP server. So basically identities are provided by LDAP.

I don't know Pomerium users but I can imagine some of them would like to use a self-hosted identity solution with their self-hosted auth proxy (Pomerium). Maybe some users have identities provided by Active Directory for example. I don't know...

One thing is sure: for all those identity providers where _allowed_groups_ is not supported, there's no right (scalable, durable) way to perform authorization. I can't imagine modifying Pomerium configuration every time there's a new/deleted user, an user changes email address or access policies have changed for an user.

PS: Would love to contribute to a solution (fallback idea seems good to me) but unfortunately I don't speak Go... :smirk:

**Comment by dmoose** [comment]
I'm looking to do the exact same thing.  Traefik, Pomerium, and Dex.  Very interested in the resolution of this since reviewing docs for Pomerium and Dex has not led me to a workable solution.

**Comment by desimone** [comment]
@dmoose What identity provider are you using with Dex?

**Comment by dmoose** [comment]
github internally, but we have a requirement for ldap which is what pointed me at dex.  Also considering fusionAuth, but ldap moves it into the pay tier and this is early in the process.  One client will not allow the system to have an internet connection so the external identity providers are not an option.

**Comment by sly-net** [comment]
@dmoose In short, on my side, I've found no solution. I'm still using Traefik but by managing authorization (with Pomerium policies) by users, instead of groups as I wished. It's far from ideal.

The biggest concerns that may apply if you let users change their email address themselves is impersonation or loss of access. Simply because when applying policies by users, access are allowed using email addresses in Pomerium. If an user changes email address, then she/he may be granted access to a service she/he couldn't access originally. Or she/he can simply loose all access if Pomerium policy configuration is not updated. Also, with LDAP, users are generally not identified by email address but by name, like _john_. In other words, email address is not the "primary key". By default, _n_ users can have the same email address.

Even if you accept managing policies by users, you'll likely face another issue: #834 This applies only if you want your application to know which user has been authenticated on Pomerium. No solution for this issue too, except an awful workaround: [retrieve the username](https://github.com/sly-net/netbox-remote-authn-ldap-authz/commit/36e6c9eff5ae149fd12facfa99ad7e2906235a28) from the HTTP header _X-Pomerium-Claim-User_. Not recommended in any case for production.

In conclusion, Pomerium is great when users are identified by email address but needs improvement when they are identified by username. I guess development efforts are mostly focused on IdP-as-a-Service providers.

**Comment by desimone** [comment]
> github internally, but we have a requirement for ldap which is what pointed me at dex. Also considering fusionAuth, but ldap moves it into the pay tier and this is early in the process. One client will not allow the system to have an internet connection so the external identity providers are not an option.

Thanks for the context @dmoose , that's helpful.

> The biggest concerns that may apply if you let users change their email address themselves is impersonation or loss of access. Simply because when applying policies by users, access are allowed using email addresses in Pomerium. If an user changes email address, then she/he may be granted access to a service she/he couldn't access originally. Or she/he can simply loose all access if Pomerium policy configuration is not updated. Also, with LDAP, users are generally not identified by email address but by name, like john. In other words, email address is not the "primary key". By default, n users can have the same email address.

Thanks @sly-net for fleshing out your use-case / issue with the current model, it's very helpful. 

In full disclosure, making Pomerium so `email` centric (instead of `sub`) was a bad early design decision on my part that [I would like to correct](https://github.com/pomerium/pomerium/issues/573). Would `sub` claim correspond to the user's actual primary key / guid? 

> No solution for this issue too, except an awful workaround: retrieve the username from the HTTP header X-Pomerium-Claim-User. Not recommended in any case for production.

Agree. We can do better here. 

>  I guess development efforts are mostly focused on Identity-as-a-Service providers

Though I can understand why it feels that way, that is not the case from my point of view. Rather, we are _initially_ focused on identity providers that support modern auth flows like OIDC (and to a lesser extent OAuth2) natively. For example, we support on premise active directory via AD FS in Windows Server. 

The challenge comes from supporting identity providers that don't integrate with those auth flow methods. My preferred solution would be to support the identity bridging capabilities directly, without having to use something like Dex. 

For the large part, using Pomerium with Dex (and other identity provider bridge products) work really well together. From Pomerium's point of view, Dex looks like any other identity provider with one big exception which is `groups` which is where the two products try to do more or less the same thing, but with different approaches. It's a bit out of scope of this issue, but we previously used the same method as Dex (add groups directly to the identity token), but found it broke customer deployments after a certain size. 

##  Tl;Dr

The proposed solution I am thinking of is as follows.

- Add a policy setting called `allowed_idp_claims` which allows operators to set a list of allowed claim values for _any claim key_ (e.g. `sub` or `groups`).  This setting will explicitly be distinct from `allowed_groups` which uses a directory pull mechanism. 
- Make any `X-Pomerium-Claim-*` signed with the same key the attestation header is. Since HTTP headers are technically a map of arrays, I think we can simply add the hash as a secondary value to the map. Or, we could do something like `X-Pomerium-Claim-*-sig` since HTTP header order is not guaranteed.

**Comment by sly-net** [comment]
Hi @desimone. Thank you for your time.

This is what the payload of the JWT looks like:
```
{
    "iss":"https://accounts.sld.tld",
    "sub":"CgRkZW1vEgRsZGFw",
    "aud":"pomerium",
    "exp":1599813909,
    "iat":1599727509,
    "at_hash":"8wxOV59QC3u6bh6vtg8zHw",
    "email":"demo@sld.tld",
    "email_verified":true,
    "groups":[
        "netbox-users",
        "netbox-readonly1"
    ],
    "name":"Firstname LASTNAME",
    "federated_claims":{
        "connector_id":"ldap",
        "user_id":"demo"
    }
}
```
The username is _demo_. So the username is accessible via _$.federated_claims.user_id_.

> Though I can understand why it feels that way, that is not the case from my point of view. Rather, we are initially focused on identity providers that support modern auth flows like OIDC (and to a lesser extent OAuth2) natively. For example, we support on premise active directory via AD FS in Windows Server.

> The challenge comes from supporting identity providers that don't integrate with those auth flow methods. My preferred solution would be to support the identity bridging capabilities directly, without having to use something like Dex.

> For the large part, using Pomerium with Dex (and other identity provider bridge products) work really well together. From Pomerium's point of view, Dex looks like any other identity provider with one big exception which is groups which is where the two products try to do more or less the same thing, but with different approaches. It's a bit out of scope of this issue, but we previously used the same method as Dex (add groups directly to the identity token), but found it broke customer deployments after a certain size.

Thanks for the information. Good to know. I'm not experienced in identity providers but if there are not many IdP that do not support modern/standard authentication flows, while being used by many people, it seems to be a good idea to support them directly in Pomerium. I think LDAP is often used for example.
However, it may not be an excellent idea actually if we think about user federation for example. I have no experience in this field but I guess in those more complicated designs, it's normal to have "connectors" because there are several source of identities (e.g. an OpenLDAP server and an Active Directory).

**Comment by servo1x** [comment]
This would be great because our org has restricted access to the user info endpoint and do not want to provide full access to reading group directory data on azure.

**Comment by desimone** [comment]
@servo1x et al,

I believe this is fixed as part of https://github.com/pomerium/pomerium/pull/1542 by @calebdoxsey
